server/src/test/java/org/elasticsearch/cluster/coordination/CoordinationStateTests.java

```diff
@@ -827,4 +839,177 @@ public ClusterState getLastAcceptedState() {
+                        .ifPresent(ac -> broadcast(message.targetNode, ac));
+                } else if (payload instanceof ApplyCommitRequest) {
+                    maybeNode.get().state.handleCommit((ApplyCommitRequest) payload);
+                } else {
+                    throw new AssertionError("unknown message type");
+                }
+            }
+        }
+
+        void runRandomly(int iterations) {
+            final long maxTerm = 4;
+            long nextTerm = 1;
+            for (int i = 0; i < iterations; i++) {
+                try {
+                    if (rarely() && nextTerm < maxTerm) {
+                        final long term = rarely() ? randomLongBetween(0, maxTerm + 1) : nextTerm++;
+                        final StartJoinRequest startJoinRequest = new StartJoinRequest(randomFrom(clusterNodes).localNode, term);
+                        broadcast(startJoinRequest.getSourceNode(), startJoinRequest);
+                    } else if (rarely()) {
+                        randomFrom(clusterNodes).setInitialState(initialConfiguration, initialValue);
+                    } else if (rarely() && rarely()) {
+                        randomFrom(clusterNodes).reboot();
+                    } else if (rarely()) {
+                        final List<ClusterNode> masterNodes = clusterNodes.stream().filter(cn -> cn.state.electionWon())
+                            .collect(Collectors.toList());
+                        if (masterNodes.isEmpty() == false) {
+                            final ClusterNode clusterNode = randomFrom(masterNodes);
+                            final long term = rarely() ? randomLongBetween(0, maxTerm + 1) : clusterNode.state.getCurrentTerm();
+                            final long version = rarely() ? randomIntBetween(0, 5) : clusterNode.state.getLastPublishedVersion() + 1;
+                            final VotingConfiguration acceptedConfig = rarely() ? randomVotingConfig() :
+                                clusterNode.state.getLastAcceptedConfiguration();
+                            final PublishRequest publishRequest = clusterNode.state.handleClientValue(
+                                clusterState(term, version, clusterNode.localNode, clusterNode.state.getLastCommittedConfiguration(),
+                                    acceptedConfig, randomLong()));
+                            broadcast(clusterNode.localNode, publishRequest);
+                        }
+                    } else if (messages.isEmpty() == false) {
+                        applyMessage(randomFrom(messages));
+                    }
+
+                    // check node invariants after each iteration
+                    clusterNodes.forEach(cn -> cn.state.invariant());
+                } catch (CoordinationStateRejectedException e) {
+                    // ignore
+                }
+            }
+
+            // check system invariants
```

Comment:
It caught me out that this can happen outside the loop, but on closer inspection the invariants checked are both monotonic so it's sufficient to check them at the end. I think this is worth a comment.

Reply:
sure, fixed in 711447b